### DISCUSSION TITLE
[Feature request]: Recursion scope adverb

### DISCUSSION THREAD

**Discussion by cchando** [jsoftware/jsource#240] [Ideas]
This would be backward compatible since no existing user code can contain <code>R.</code> (or whatever new adverb spelling/inflection is chosen), as it would've been a spelling error. 

The following provides a motivation:
 <pre>   ap  =: {{y`:6>x}}`{{y`:6&>/x}}@.(2=#@[)
   r   =: ($: 2&}.)`(ap 1&{)@.(ap {.)`ap@.(1=#@])
   Cond=: {{< :(,&<) r m"_}}</pre>

Here the recursive subexpression <code>r</code> is defined on its own line just because we don't want the <code>< :(,&<)</code> part of <code>Cond</code> to be part of the recursion; but <code>r</code> is not a very natural expression to separate from <code>Cond</code> and assign to a name.

Ideally this would be written inline using a recursion scoping adverb <code>R.</code>: 

 <pre>   ap  =: {{y`:6>x}}`{{y`:6&>/x}}@.(2=#@[)
   Cond=: {{< :(,&<) ($: 2&}.)`(ap 1&{)@.(ap {.)`ap@.(1=#@])R. m"_}}</pre>

where the scope of <code>$:</code> is limited to [the verb operand to the nearest adverb <code>R.</code>] which contains <code>$:</code>. The existing definition of <code>$:</code> holds if its no <code>R.</code> is present within the largest verb containing <code>$:</code> within the given sentence. <code>R.</code> can appear in multiple places within a verb, as in 
<pre>   (… $: … ($: …)R. …)R.</pre>

As motivation, one of the benefits of J's terseness, and of tacit programming's absence of names, is the dispensation of the need to break up code into arbitrary (non-[conceptual/semantic]) pieces which will be used once only, and of the resulting need to create arbitrary names for these pieces. Anonymous recursion seems to be the only place where one is forced into doing so. 

<code>Z.</code> would be my fallback naming suggestion since this adverb is like <code>Z:</code> in that each modifies the behavior of a particular primitive (<code>$:</code> and <code>F(.:)(.:)</code> respectively).

**Comment by HenryHRich** [comment]
I don't see any objection
